### PR TITLE
[WIP]add data processing

### PR DESCRIPTION
**Pull request by yiyixuxu**
I created a section about data in README, but specifically, this PR added a section "Prepare a large local dataset" with relevant scripts.

### CODE REVIEW

File: jax-controlnet-sprint/coyo_1m_dataset_preprocess.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import requests
+import numpy as np
+import random
+import jsonlines
+
+import logging
+logger = logging.getLogger(__name__)
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="example of a data preprocessing script.")
```
Comment (sayakpaul) [inline jax-controlnet-sprint/coyo_1m_dataset_preprocess.py:15]:
```suggestion
    parser = argparse.ArgumentParser(description="Example of a data preprocessing script.")
```

Nit.

File: jax-controlnet-sprint/README.md [hunk 2]
```diff
@@ -74,3 +75,102 @@ We will have talks from folks working at JAX & TPU teams at Google, diffusers an
...
+
+#### Mount a disk 
+
+If you need extra space, you can follow [this guide](https://cloud.google.com/tpu/docs/setup-persistent-disk#prerequisites) to create a persistent disk, attach it to your TPU VM, and create a directory to mount the disk. You can then use this directory to store your dataset.
+
+#### Data preprocessing 
+
+Here we demonstrate how to prepare a large dataset to train a ControlNet model with canny edge detection. More specifically, we provide an [example script](./coyo_1m_dataset_preprocess.py) that:
+* Select 1 million image-text pairs from an existing dataset [COYO-700M](https://huggingface.co/datasets/kakaobrain/coyo-700m)
+* Download each image and use Canny edge detector to generate the conditioning image.
```
Comment (sayakpaul) [inline jax-controlnet-sprint/README.md:93]: What conditioning image will be generated? Pose, Canny? Give some hint about that here.

File: jax-controlnet-sprint/README.md [hunk 3]
```diff
@@ -74,3 +75,102 @@ We will have talks from folks working at JAX & TPU teams at Google, diffusers an
...
+### Prepare a large local dataset
+
+#### Mount a disk 
+
+If you need extra space, you can follow [this guide](https://cloud.google.com/tpu/docs/setup-persistent-disk#prerequisites) to create a persistent disk, attach it to your TPU VM, and create a directory to mount the disk. You can then use this directory to store your dataset.
+
+#### Data preprocessing 
+
+Here we demonstrate how to prepare a large dataset to train a ControlNet model with canny edge detection. More specifically, we provide an [example script](./coyo_1m_dataset_preprocess.py) that:
+* Select 1 million image-text pairs from an existing dataset [COYO-700M](https://huggingface.co/datasets/kakaobrain/coyo-700m)
```
Comment (sayakpaul) [inline jax-controlnet-sprint/README.md:92]:
```suggestion
* Selects 1 million image-text pairs from an existing dataset [COYO-700M](https://huggingface.co/datasets/kakaobrain/coyo-700m).
```

File: jax-controlnet-sprint/README.md (hunk 2 above)
Comment (sayakpaul) [reply jax-controlnet-sprint/README.md:93]:
```suggestion
* Downloads each image and use Canny edge detector to generate the conditioning image. 
```

File: jax-controlnet-sprint/README.md [hunk 4]
```diff
@@ -74,3 +75,102 @@ We will have talks from folks working at JAX & TPU teams at Google, diffusers an
...
+│   ├── .......
+│   └── image_1000000.jpeg
+├── processed_images
+│   ├── image_1.png
+│   ├── .......
+│   └── image_1000000.jpeg
+└── meta.jsonl
+```
+
+#### load dataset
```
Comment (sayakpaul) [inline jax-controlnet-sprint/README.md:121]:
```suggestion
#### Load dataset
```

File: jax-controlnet-sprint/coyo_1m_dataset_preprocess.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+if __name__ == "__main__":
+
+    args = parse_args()
+
+    # load coyo-700
+    dataset = load_dataset(
+        "kakaobrain/coyo-700m",
+        cache_dir=args.cache_dir,
+        split='train',
+        )
```
Comment (sayakpaul) [inline jax-controlnet-sprint/coyo_1m_dataset_preprocess.py:69]: Could we please format the *.py scripts with `black`, `isort`, and `ruff`? We can use the version that are in `diffusers`.

File: jax-controlnet-sprint/README.md [hunk 6]
```diff
@@ -74,3 +75,102 @@ We will have talks from folks working at JAX & TPU teams at Google, diffusers an
...
+
+To load a dataset from the data folder you just created, you should add a dataset loading script to your data folder. The dataset loading script should have the same name as the folder. For example, if your data folder is `data`, you should add a data loading script named `data.py`. We provided an [example data loading script](./data.py) for you to use. All you need to do is to update the `DATA_DIR` with the correct path to your data folder. For more details about how to write a dataset loading script, refer to the [documentation](https://huggingface.co/docs/datasets/dataset_script)
```
Comment (sayakpaul) [inline jax-controlnet-sprint/README.md:123]:
```suggestion
To load a dataset from the data folder you just created, you should add a dataset loading script to your data folder. The dataset loading script should have the same name as the folder. For example, if your data folder is `data`, you should add a data loading script named `data.py`. We provide an [example data loading script](./data.py) for you to use. All you need to do is to update the `DATA_DIR` with the correct path to your data folder. For more details about how to write a dataset loading script, refer to the [documentation](https://huggingface.co/docs/datasets/dataset_script).
```

File: jax-controlnet-sprint/README.md [hunk 7]
```diff
@@ -74,3 +75,102 @@ We will have talks from folks working at JAX & TPU teams at Google, diffusers an
...
+
+
+
+
+
+
+
+
+
+
```
Comment (sayakpaul) [inline jax-controlnet-sprint/README.md:176]: Can we delete these blank spaces?

**Review by sayakpaul — APPROVED** [review]
@yiyixuxu thanks a lot for working on this utility! It will be very useful. My points are mainly about nits. 

Additionally, I have the following suggestions:

* Should we add these scripts to a new folder called "data_processing"? That way, the [Colab](https://github.com/huggingface/community-events/pull/140) being added in https://github.com/huggingface/community-events/pull/140 can also go to the directory. I think it will be easier for the participants to navigate. 
*

File: jax-controlnet-sprint/coyo_1m_dataset_preprocess.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        try:
+            # download original image
+            image = Image.open(requests.get(image_url, stream=True, timeout=5).raw)
+            image_path = f"{args.train_data_dir}/images/{example['id']}.png"
+            image.save(image_path)
+        
+            # generate and save canny image
+            processed_image= np.array(image)
+            threholds = (random.randint(0,255), random.randint(0,255))  # generate random threholds 
+            processed_image= cv2.Canny(processed_image, min(threholds), max(threholds))
```
Comment (sayakpaul) [inline jax-controlnet-sprint/coyo_1m_dataset_preprocess.py:90]: Do we need to convert to RGB -> BGR and BGR -> RGB again before serialization since we don't use `cv2.imwrite`?

File: jax-controlnet-sprint/coyo_1m_dataset_preprocess.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            # write to meta.jsonl
+            meta = {'image': image_path, 'conditioning_image': processed_image_path, 'caption': example['text']}
+            with jsonlines.open(f"{args.train_data_dir}/meta.jsonl", "a") as writer:   # for writing
+                writer.write(meta)
+
+        except Exception as e:
+            logger.error(f"Failed to process image{image_url}: {str(e)}")
+
+    # preprocess -> image, processed image and meta.jsonl
+    small_dataset.map(preprocess_and_save, num_proc=args.num_proc)
```
Comment (sayakpaul) [inline jax-controlnet-sprint/coyo_1m_dataset_preprocess.py:106]: Might not be a small dataset, though hahaha.

**Comment by yiyixuxu** [comment]
> * Should we add these scripts to a new folder called "data_processing"? That way, the [Colab](https://github.com/huggingface/community-events/pull/140) being added in [[JAX/Diffusers Sprint] adds a dataset to show an end-to-end workflow of dataset creation. #140](https://github.com/huggingface/community-events/pull/140) can also go to the directory. I think it will be easier for the participants to navigate.

Yeah, I think a new folder makes sense since we have quite a lot of content around data_processing 
The media pipe notebook is great! It provides an in-depth tutorial, I think we should add a section in the README to talk about it; also probably need to talk about loading datasets in streaming mode, and mention the ready-to-use datasets @merveenoyan is uploading.

**Comment by sayakpaul** [comment]
Once the contents are in, I can update the README.

File: jax-controlnet-sprint/coyo_1m_dataset_preprocess.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                random.randint(0, 255),
+            )
+            processed_image = cv2.Canny(processed_image, min(threholds), max(threholds))
+            processed_image = processed_image[:, :, None]
+            processed_image = np.concatenate(
+                [processed_image, processed_image, processed_image], axis=2
+            )
+            processed_image = Image.fromarray(processed_image)
+            processed_image_path = (
+                f"{args.train_data_dir}/processed_images/{example['id']}.png"
```
Comment (sayakpaul) [inline jax-controlnet-sprint/coyo_1m_dataset_preprocess.py:110]: I think `conditioning_images` is a better prefix for saving the conditioning images?

**Review by sayakpaul — APPROVED** [review]
I think we should be good to merge this one. @merveenoyan could you also take a look?

**Comment by sayakpaul** [comment]
@yiyixuxu I just merged #140. 

The [notebook](https://github.com/huggingface/community-events/blob/main/jax-controlnet-sprint/dataset_tools/create_pose_dataset.ipynb) now lies here: https://github.com/huggingface/community-events/tree/main/jax-controlnet-sprint/dataset_tools.

Could you 

* Pull the changes from `main` in your branch? 
* Move the scripts you're introducing in this PR to `dataset_tools`? 
* I am writing some suggestions to the README to include a note on the notebook as well.

File: jax-controlnet-sprint/README.md [hunk 11]
```diff
@@ -74,3 +75,89 @@ We will have talks from folks working at JAX & TPU teams at Google, diffusers an
...
+ --cache_dir=$DISK_DIR \
+ --resolution=512 \
+ --learning_rate=1e-5 \
+ --train_batch_size=2 \
+ --revision="non-ema" \
+ --from_pt \
+ --max_train_steps=500000 \
+ --checkpointing_steps=10000 \
+ --dataloader_num_workers=16 
+ ```
```
Comment (sayakpaul) [inline jax-controlnet-sprint/README.md:163]:
```suggestion

### Prepare a dataset with MediaPipe and Hugging Face 

We provide a notebook ([![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/community-events/blob/main/jax-controlnet-sprint/dataset_tools/create_pose_dataset.ipynb)) that shows you how to prepare a dataset for ControlNet training using [MediaPipe](https://developers.google.com/mediapipe) and Hugging Face. Specifically, in the notebook, we show:

* How to leverage MediaPipe solutions to extract pose body joints from the input images.
* Predict captions using BLIP-2 from the input images using 🤗 Transformers.
* Build and push the final dataset to the Hugging Face Hub using 🤗 Datasets. 

You can refer to the notebook to create your own datasets using other MediaPipe solutions as well. Below, we list all the relevant ones:

* [Pose Landmark Detection](https://developers.google.com/mediapipe/solutions/vision/pose_landmarker)
* [Face Landmark Detection](https://developers.google.com/mediapipe/solutions/vision/face_landmarker)
* [Selfie Segmentation](https://developers.google.com/mediapipe/solutions/vision/image_segmenter)

**Review by sayakpaul — APPROVED** [review]
Looks really good to me. 

@osanseviero or @merveenoyan could either of you take a quick look?
